File: src/search/MyPow.java
package search;

/**
 * 实现 pow(x, n) ，即计算 x 的 n 次幂函数。
 */
public class MyPow {
    public double myPow(double x, int n) {
        if (n < 0) return 1 / power(x, -n);
        return power(x, n);
    }
    private double power(double x, int n) {
        if (n == 0) return 1;
        double half = power(x, n / 2);
        if (n % 2 == 0) return half * half;
        return x * half * half;
    }
    public static void main(String[] args){
        System.out.println(new MyPow().myPow(8.84372,5));
    }
}

File: src/Mypow.kt
import com.sun.xml.internal.fastinfoset.util.StringArray

class Mypow {
    fun myPow(x: Double, n: Int): Double {
        if (n == 0) return 1.0;
        if(n==Int.MIN_VALUE){//Int最小值取反超出int上限
            return 1 / (pow(x, -(n+2))*x*x)
        }
        return if (n < 0) 1 / pow(x, -n) else pow(x, n);
    }

    private fun pow(x: Double, n: Int): Double {
        println(n);
        if (n == 1) return x;
        val tmp = pow(x, n / 2);
        return if (n % 2 == 0) tmp * tmp else tmp * tmp * x;
    }

    companion object {
        @JvmStatic
        public fun main(args: Array<String>) {
           // print(Mypow().myPow(1.0, -2147483648))
            System.out.println(Int.MIN_VALUE)
            System.out.println(-Int.MIN_VALUE)
            System.out.println(Int.MAX_VALUE)
            System.out.println(-Int.MAX_VALUE)
        }
    }
}